streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -5715,37 +5716,73 @@ object ZStream extends ZStreamPlatformSpecificConstructors {
   }
 
   private[zio] class Rechunker[A](n: Int) {
-    private var buffer: ChunkBuilder[A] = if (n > 1) ChunkBuilder.make(n) else null
-    private var pos: Int                = 0
+    private var buffer: Chunk[A]              = if (n > 1) Chunk.empty[A] else null
+    private var chunkBuilder: ChunkBuilder[A] = null
+    private var pos: Int                      = 0
 
     def isEmpty: Boolean = pos == 0
 
     final def rechunk(
       chunk: Chunk[A]
     )(implicit trace: Trace): ZChannel[Any, ZNothing, Any, Any, ZNothing, Chunk[A], Any] = {
-      val len = chunk.size
-      if (len == 0) {
+      val chunkSize = chunk.size
+      if (chunkSize == 0) {
         null
-      } else if (isEmpty && len == n) {
+      } else if (isEmpty && chunkSize == n) {
         ZChannel.write(chunk)
       } else if (n == 1) {
-        rechunk1(chunk, len)
-      } else {
+        rechunk1(chunk, chunkSize)
+      } else if (chunkSize < 22) {
+        // Optimized for small chunks. The limit 22 comes from testing with benchmarks
         var i = 0
 
         var channel: ZChannel[Any, ZNothing, Any, Any, ZNothing, Chunk[A], Any] = null
+        if (chunkBuilder == null) chunkBuilder = ChunkBuilder.make(chunkSize - pos)
 
-        while (i < len) {
-          buffer += chunk(i)
+        while (i < chunkSize) {
+          chunkBuilder += chunk(i)
           i += 1
           pos += 1
           if (pos == n) {
-            val bufferResult = ZChannel.write(buffer.result())
-            channel =
-              if (channel eq null) bufferResult
-              else channel *> bufferResult
+            buffer ++= chunkBuilder.result()
+            chunkBuilder = ChunkBuilder.make(chunkSize - pos)
```

Comment:
Maybe I'm missing something, but doesn't calling `.clear()` on the existing builder to avoid creating a new one works as well? You won't have to size the new builder either because it'll reuse the existing arrays